app/models/manageiq/providers/physical_infra_manager.rb

```diff
@@ -110,6 +110,10 @@ def console_url
       raise MiqException::Error, _("Console not supported")
     end
 
+    def self.firmware_update_class
+      self::FirmwareUpdate
```

Comment:
I only found the `FirewareUpdate` class defined in the Redfish PR https://github.com/ManageIQ/manageiq-providers-redfish/pull/69/files.  I think we should have a `FirmwareUpdate` class at this intermediate class level and then the Provider class is sub-classed from this new one.  

Something like:
```ruby
class ManageIQ::Providers::PhysicalInfraManager::FirmwareUpdate < ::PhysicalServerFirmwareUpdateTask
  # Firmware update is triggered in a single Redfish API call for a list of servers
  # hence we don't need more than one task per firmware update request.
  SINGLE_TASK = true

  include_concern 'StateMachine'
end
```

Then the RedFish `FirmwareUpdate` class is derived from this new intermediate class.

Also, since this is a task is there a reason you picked the naming `FirmwareUpdate`?  The name sounds more like an inventoried item then a task that would run a state-machine.

Reply:
Hi @gmcculloug and sorry for delayed response due to my annual PTO. I've applied both your comments, the reason it was implemented different before is that I took VM provision request as a guideline where they

a) implement but don't ever extend the intermediate class (https://github.com/ManageIQ/manageiq/blob/master/app/models/manageiq/providers/cloud_manager/provision.rb#L1), see how providers then extend the Task directly anyway: https://github.com/ManageIQ/manageiq-providers-vmware/blob/master/app/models/manageiq/providers/vmware/infra_manager/provision.rb#L1

b) name the task as `Provision` and not as `ProvisionTask` https://github.com/ManageIQ/manageiq-providers-vmware/blob/master/app/models/manageiq/providers/vmware/infra_manager/provision.rb#L1

But I'm with you on this one, because seeing 'Provision' or 'FirmwareUpdate' inside provider feels like inventory item.

Reply:
Thanks @miha-plesko.  The `MiqProvision` class is a little special as it existed before we combined the request and tasks in sub-classed models.  The name was kept for backward compatibility.

In newer models we use request/task as the model suffix.